docs: fix description of treemanager api (#1135)
Refinitiv/refinitiv-ui: documents/src/pages/elements/tree-select.md

This control will abort any current selection changes and go back to the tree state when first opened.

## Manipulating item properties

> Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`.

Reviewer: ```suggestion
Item properties of Tree Select could be read and updated programmatically through its [Tree Manager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`.
```

Updating item's `selected` would not affected TreeSelect's `values` as the selection needs to be confirmed first. This could be done by using "Done" button or calling `commit()` method.

```javascript
// Select the item which value is 'ADR'
const treeSelect = document.querySelector('ef-tree-select');
treeSelect.data = [
  {
    label: 'Algeria',
    value: 'DZA',
    items: [{
      value: 'ADR',
      label: 'Adrar'
    }, {
      value: 'TAM',
      label: 'Tamanghasset'
    }]
  }
];
console.log(treeSelect.values); // Expected output: []

const treeNodes = treeSelect.treeManager.getTreeNodes();

> const selectedNode = treeNodes.find(treeNode => treeNode.value === 'ADR');

Reviewer: `selectedNode` is rather misleading: the node is not selected yet.
```suggestion
const node = treeNodes.find(treeNode => treeNode.value === 'ADR');
```

> selectedNode.selected = true; // Current output: []

Reviewer: ```suggestion
selectedNode.selected = true;
```

treeSelect.commit();
console.log(treeSelect.values); // Expected output: ['ADR']
```

```typescript
// Select the item which value is 'ADR'
const treeSelect = document.querySelector('ef-tree-select')!;

> treeSelect.data = [

Reviewer: `data` would be an unknown property without proper type def. Missing import type { TreeSelect } from '@refinitiv-ui/elements/tree-select'?

Author: Fixed.

  {
    label: 'Algeria',
    value: 'DZA',
    items: [{
      value: 'ADR',
      label: 'Adrar'
    }, {
      value: 'TAM',
      label: 'Tamanghasset'
    }]
  }
];
console.log(treeSelect.values); // Expected output: []

const treeNodes = treeSelect.treeManager.getTreeNodes();

> const selectedNode = treeNodes.find(treeNode => treeNode.value === 'ADR');

Reviewer: `selectedNode` is rather misleading: the node is not selected yet.
```suggestion
const node = treeNodes.find(treeNode => treeNode.value === 'ADR');
```

> selectedNode.selected = true; // Current output: []

Reviewer: ```suggestion
selectedNode.selected = true;
```

treeSelect.commit();
console.log(treeSelect.values); // Expected output: ['ADR']
```

## Custom renderer

> TreeSelect supports custom rendering by providing a renderer property that can define how each item is displayed. The renderer is a callback function that takes a data item, a Collection Composer, and previously mapped item elements (if any), and must return an HTMLElement. You can use TreeNode to provides access to the current item's data property. Note that for performance sensitive use cases such as a large number of items, consider using [Collection Composer](/custom-components/utils/data-management#collection-composer) instead.

Reviewer: ```suggestion
Tree Select defines how each of its item is displayed with `renderer` property. You can customise this renderer by setting a callback function to the property. [Tree Node](/custom-components/utils/tree-node) is the easiest way to implement the function. Note that for performance sensitive use cases such as a large number of items, consider using [Collection Composer](/custom-components/utils/data-management#collection-composer) instead.
```

```javascript
import { uuid } from '@refinitiv-ui/utils/uuid.js';
import { CheckedState, TreeManager, TreeManagerMode } from '@refinitiv-ui/elements/tree';

> // Implement TreeSelect's default render with Tree Node instead of Collection Composer

Reviewer: ```suggestion
// Implement Tree Select's default renderer with Tree Node instead of Collection Composer
```

// for comparison, check https://github.com/Refinitiv/refinitiv-ui/blob/v7/packages/elements/src/tree/helpers/renderer.ts
const createTreeRenderer = (context) => {
  const key = uuid();

  return (item, composer, element = document.createElement('ef-tree-item')) => {
    const multiple = context?.multiple === true;
    const noRelation = context?.noRelation === true;
    const mode = !multiple || !noRelation ? TreeManagerMode.RELATIONAL : TreeManagerMode.INDEPENDENT;
    const manager = context?.manager || context?.treeManager || new TreeManager(composer, mode);

    const treeNode = manager.getTreeNode(item);
    element.multiple = multiple;
    element.item = item;
    element.id = `${key}-${item.value || ''}`;
    element.depth = treeNode.getDepth();
    element.parent = treeNode.isParent();
    element.expanded = treeNode.expanded;
    element.checkedState =
      !multiple && element.parent ? CheckedState.UNCHECKED : treeNode.getCheckedState();
    element.icon = treeNode.icon;
    element.label = treeNode.label;
    element.disabled = treeNode.disabled;
    element.readonly = treeNode.readonly;
    element.highlighted = treeNode.highlighted;

    return element;
  };
};

> treeSelect.renderer = createTreeRenderer(this)

Reviewer: ```suggestion
treeSelect.renderer = createTreeRenderer(treeSelect)
```

```

```typescript
import { CheckedState, TreeManager, TreeManagerMode } from '@refinitiv-ui/elements/tree';
import { uuid } from '@refinitiv-ui/utils/uuid.js';
import type { CollectionComposer } from '@refinitiv-ui/utils/collection.js';
import type { TreeSelectDataItem, TreeSelectItem } from '@refinitiv-ui/elements/tree-select';

type RendererScope<T extends TreeSelectDataItem> = {
  multiple?: boolean;
  noRelation?: boolean;
  manager?: TreeManager<T>;
  treeManager?: TreeManager<T>;
};

> // Implement TreeSelect's default render with Tree Node instead of Collection Composer

Reviewer: ```suggestion
// Implement Tree Select's default renderer with Tree Node instead of Collection Composer
```

// for comparison, check https://github.com/Refinitiv/refinitiv-ui/blob/v7/packages/elements/src/tree/helpers/renderer.ts
export const createTreeRenderer = <T extends TreeSelectDataItem = TreeSelectDataItem>(
  context?: unknown
): ((item: T, composer: CollectionComposer<T>, element?: HTMLElement) => HTMLElement) => {
  const key: string = uuid();

  return (
    item: T,
    composer: CollectionComposer<T>,
    element: HTMLElement = document.createElement('ef-tree-item')
  ): HTMLElement => {
    const _context = context as RendererScope<T> | undefined;
    const _element = element as TreeSelectItem;
    const multiple = _context?.multiple === true;
    const noRelation = _context?.noRelation === true;
    const mode = !multiple || !noRelation ? TreeManagerMode.RELATIONAL : TreeManagerMode.INDEPENDENT;

>     const manager = _context?.treeManager || new TreeManager(composer, mode);

Reviewer: ```suggestion
    const manager = _context?.manager || _context?.treeManager || new TreeManager(composer, mode);
```

    const treeNode = manager.getTreeNode(item);
    _element.multiple = multiple;
    _element.item = item;
    _element.id = `${key}-${item.value || ''}`;
    _element.depth = treeNode.getDepth();
    _element.parent = treeNode.isParent();
    _element.expanded = treeNode.expanded;
    _element.checkedState =
      !multiple && _element.parent ? CheckedState.UNCHECKED : treeNode.getCheckedState();
    _element.icon = treeNode.icon;
    _element.label = treeNode.label;
    _element.disabled = treeNode.disabled;
    _element.readonly = treeNode.readonly;
    _element.highlighted = treeNode.highlighted;

    return _element;
  };
};

> treeSelect.renderer = createTreeRenderer(this)

Reviewer: ```suggestion
treeSelect.renderer = createTreeRenderer(treeSelect)
```

This control will abort any current selection changes and go back to the tree state when first opened.

> ## Manipulate item property

Reviewer: ```suggestion
## Manipulating item properties
```

> Tree Select component supports functionality for updating and reading data to manipulate items properties. These are used in conjunction with [TreeManager](./custom-components/utils/tree-manager) from `treeManager` property and [TreeNode](./custom-components/utils/tree-node) which can be retrieved from `getTreeNode` through `treeManager` property.

Reviewer: ```suggestion
Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. 
```

```javascript

> const treeSelect = document.querySelector('ef-tree-select');

Reviewer: let's provide a more concrete example. Check Tree Node doc in `feat/tree-node/docs-tree-node` for a real example.

Author: Added.

const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
node0.label = 'new label'; // edit label property which this will reflect to the component.
```

```typescript
const treeSelect = document.querySelector('ef-tree-select');
if (treeSelect) {
  const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
  node0.label = 'new label'; // edit label property which this will reflect to the component.

> }

Reviewer: non-null assertion should be typescript example easier to read.

```suggestion
const treeSelect = document.querySelector('ef-tree-select')!;
const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
node0.label = 'new label'; // edit label property which this will reflect to the component.
```

This control will abort any current selection changes and go back to the tree state when first opened.

## Manipulating item properties

> Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. 

Reviewer: let's cleanup trailing whitespaces.

> Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. 

Reviewer: ```suggestion
Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In adiition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too. 
```

> Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In adiition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.

Reviewer: ```suggestion
Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.
```

> Item properties of Tree Select could be read and updated programmatically through its [TreeManager](./custom-components/utils/tree-manager) which is available as `treeManager` property. Retrieve [TreeNode(s)](./custom-components/utils/tree-node) representing each item by calling `getTreeNode()` or `getTreeNodes()` of `treeManager`. In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.

Reviewer 2: `In addition, they could be used for [custom renderer](./custom-components/utils/tree-node#custom-renderer) too.`

I feel that this content should be removed from this section. What we could do is to introduce another section, 'Custom Renderer' so we can provide completed guideline.

Reviewer: Let's move "Custom Renderer" from "Tree Node" to "Tree" & "Tree Select" + revise it content a little as readers focus on Component first rather than "Tree Node" first.

Author: Moved and wrote new Custom renderer section on both components.

```javascript
const treeSelect = document.querySelector('ef-tree-select');
treeSelect.data = [
  {
    label: 'Algeria',
    value: 'DZA',
    items: [{
      value: 'ADR',
      label: 'Adrar'
    }, {
      value: 'TAM',
      label: 'Tamanghasset'
    }]
  }
];

> const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.

Reviewer: naming variable with numbers looks sloppy.

Author: Fixed.

> const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.

Author: ```suggestion
const firstNode = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```

node0.label = 'DZ'; // edit label property which this will reflect to the component.
```

```typescript
const treeSelect = document.querySelector('ef-tree-select')!;
treeSelect.data = [
  {
    label: 'Algeria',
    value: 'DZA',
    items: [{
      value: 'ADR',
      label: 'Adrar'
    }, {
      value: 'TAM',
      label: 'Tamanghasset'
    }]
  }
];

> const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.

Reviewer: naming variable with numbers looks sloppy.

Author: Fixed.

> const node0 = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.

Author: ```suggestion
const firstNode = treeSelect.treeManager.getTreeNodes()[0]; // get the item at index number 0.
```
